python_modules/libraries/dagster-aws/dagster_aws/emr/monitor.py

```diff
@@ -0,0 +1,185 @@
+"""Monitoring for EMR on EKS job runs."""
+import base64
+import itertools
+import json
+import logging
+import pickle
+import sys
+import time
+from dataclasses import dataclass
+from logging import Logger
+from typing import Iterator, List
+
+import boto3
+from dagster.core.events import DagsterEvent
+
+from .cloud_watch_logs_follower import CloudwatchLogsFollower
+
+STATES_ONGOING = {"PENDING", "SUBMITTED", "RUNNING"}
+STATES_SUCCESS = {"COMPLETED"}
+
+
+def _is_final_event(event: DagsterEvent) -> bool:
+    """Checks whether a Dagster event signals step success or failure"""
+    return event.is_step_failure or event.is_step_success
+
+
+@dataclass
+class EmrEksJobError(Exception):
+    """Error description for an EMR on EKS Job Run."""
+
+    failure_reason: str
+    state: str
+    state_details: str
+
+
+@dataclass
+class EmrEksJobRunMonitor:
+    """Monitor status and logs of an EMR on EKS Job Run.
+
+    The status of the Job Run is periodically queried from
+    the EMR on EKS API.
+
+    In addition, the log streams for the Spark driver are
+    tailed so that logs and events can be reported to Dagster.
+    See ``EmrEksPySparkResource`` for details.
+    """
+
+    cluster_id: str
+    region_name: str
+
+    def __post_init__(self):
+        self.emr_client = boto3.client("emr-containers", region_name=self.region_name)
+
+    def consume_tailers(
+        self,
+        log: Logger,
+        tailers: List[CloudwatchLogsFollower],
+        log_stream_name_prefix: str,
+    ) -> Iterator[DagsterEvent]:
+        """Check logs from a list of log tailers. Yield events for logs corresponding to
+        Dagster events, otherwise propagate the logs
+        """
+        # Get logs and events
+        iterators = [tailer.fetch_events(log) for tailer in tailers]
+        for log_record in itertools.chain(*iterators):
+            msg = self._try_unpack_log_record(log_record["message"])
+
+            # Remove the log stream name prefix: e.g.,
+            # `some/prefix/stdout` -> `stdout`.
+            log_stream = log_record["logStreamName"][len(log_stream_name_prefix) :]
+
+            if self._is_dagster_event(msg):
+                yield from self._yield_event(msg)
+            else:
+                self._write_log(log_stream, msg)
+
+    def wait_for_completion(
+        self,
+        log: Logger,
+        job_id: str,
+        log_group_name: str,
+        log_stream_name_prefix: str,
+        start_timestamp_ms: int,
+        wait_interval_secs: int = 10,
+        max_wait_after_done_secs: int = 120,
+    ) -> Iterator[DagsterEvent]:
```

Comment:
nit: we might need to document the parameters carefully, especially `max_wait_after_done_secs`